Improve sorting performance

igitur opened the pull request:
~Depends on #1647~

This PR implements custom `IComparer` instances to sort rows and columns instead of trying to reinvent the wheel. It's now much faster and a small bug has also been fixed (hence the updated reference files).

igitur commented:
I should probably squash the 3 "CodeMaid" commits.

igitur commented:
Since this builds on #1647, which is now merged, I'd like to merge this one next. So when you have time, @Pankraty , please review this one next.

Pankraty commented on ClosedXML/Excel/Ranges/XLRangeRowComparer.cs:
```diff
@@ -0,0 +1,81 @@
+                else
+                {
+                    if (thisCell.DataType == otherCell.DataType)
+                    {
+                        switch (thisCell.DataType)
+                        {
+                            case XLDataType.Text:
+                                comparison = e.MatchCase
+                                                 ? thisCell.InnerText.CompareTo(otherCell.InnerText)
+                                                 : String.Compare(thisCell.InnerText, otherCell.InnerText, true);
+                                break;
+
+                            case XLDataType.TimeSpan:
+                                comparison = thisCell.GetTimeSpan().CompareTo(otherCell.GetTimeSpan());
+                                break;
+
+                            case XLDataType.DateTime:
+                                comparison = thisCell.GetDateTime().CompareTo(otherCell.GetDateTime());
+                                break;
+
+                            case XLDataType.Number:
+                                comparison = thisCell.GetDouble().CompareTo(otherCell.GetDouble());
+                                break;
+
+                            case XLDataType.Boolean:
+                                comparison = thisCell.GetBoolean().CompareTo(otherCell.GetBoolean());
+                                break;
+
+                            default:
+                                throw new NotImplementedException();
+                        }
+                    }
+                    else if (e.MatchCase)
+                        comparison = String.Compare(thisCell.GetString(), otherCell.GetString(), true);
+                    else
+                        comparison = thisCell.GetString().CompareTo(otherCell.GetString());
+                }
+
+                if (comparison != 0)
+                    return e.SortOrder == XLSortOrder.Ascending ? comparison : -comparison;
```
Can't we move this into a new `XLCellComparer` and reuse it in `XLRangeRowComparer` and `XLRangeColumnComparer`? They are nearly identical.

igitur replied:
Done.

igitur commented on ClosedXML/Excel/Ranges/XLRangeColumnComparer.cs:
```diff
@@ -0,0 +1,64 @@
+                {
+                    if (thisCell.DataType == otherCell.DataType)
+                    {
+                        if (thisCell.DataType == XLDataType.Text)
+                        {
+                            comparison = e.MatchCase
+                                             ? thisCell.InnerText.CompareTo(otherCell.InnerText)
+                                             : String.Compare(thisCell.InnerText, otherCell.InnerText, true);
+                        }
```
Yes, I think this requires a separate PR, to be done first before continuing this PR.

Pankraty replied:
It can go after too, because may also require adjusting formulas on cells movement (which you implemented in this PR).

igitur replied:
Ok, let's park it. 

In fact, I dislike the `InnerText` property and `string _cellValue` field. I don't really see its purpose. I think using `.Value` or `.CachedValue` should cover all cases. Or `.GetString()` as you propose. And I think `string _cellValue` should be replaced with `object _cellValue` to avoid the conversion to `string` values in `XLCell.SetKnownTypedValue`.

igitur commented:
@Pankraty Good catches, as always. I will address the rest later. Thanks.

igitur commented on ClosedXML/Excel/Cells/XLCellsCollection.cs:
```diff
@@ -414,5 +424,90 @@ private static void IncrementUsage(Dictionary<int, int> dictionary, Int32 key)
+                }
+            });
+            return anyChanged;
+        }
```
What's the benefit of this?

igitur replied:
I just in-lined that code into the outer method. No need for a separate method at all.

igitur commented on ClosedXML/Excel/Ranges/IXLRangeColumns.cs:
```diff
@@ -39,11 +39,29 @@ public interface IXLRangeColumns : IEnumerable<IXLRangeColumn>
+        /// Returns a contiguous set of range columns
+        /// </summary>
+        /// <returns></returns>
+        IXLRangeColumns Contiguous();
```
I also see now that `Contiguous` has a few different, but related meanings.

From https://www.merriam-webster.com/dictionary/contiguous 👍 

![image](https://user-images.githubusercontent.com/145854/121803010-bebd1200-cc3f-11eb-8039-1e52f20a4f51.png)

My intention was to use the last definition `touching connected throughout in an unbroken sequence`, (almost like `continuous`) which is different to the meaning of `adjacent`.

How about if I keep the method name, but update the doctext to:
```c#
        /// <summary>
        /// Returns a set of range columns that are all touching or connected throughout in an unbroken sequence.
        /// </summary>
        IXLRangeColumns Contiguous();
```

igitur commented on ClosedXML.Tests/Excel/Ranges/SortTests.cs:
```diff
@@ -0,0 +1,90 @@
+
+            var autoFilter = ws.Range(1, 1, 14001, 2).SetAutoFilter();
+
+            var stopWatch = new System.Diagnostics.Stopwatch();
```
Fixed

igitur commented on ClosedXML/Excel/Cells/XLCellsCollection.cs:
```diff
@@ -317,105 +383,131 @@ internal void SwapRanges(XLSheetRange sheetRange1, XLSheetRange sheetRange2, XLW
-                    //if (cell1 != null)
-                    //{
+                    // Force evaluation of FormulaR1C1 and clear FormulaA1 - this will preserve the relative formula
+                    if (cell1.HasFormula && !string.IsNullOrWhiteSpace(cell1.FormulaR1C1)) cell1.ClearFormulaA1();
```
Fixed

igitur commented on ClosedXML/Excel/Cells/XLCellsCollection.cs:
```diff
@@ -317,105 +383,131 @@ internal void SwapRanges(XLSheetRange sheetRange1, XLSheetRange sheetRange2, XLW
-                    //if (cell2 == null) continue;
 
+                    // Force evaluation of FormulaR1C1 and clear FormulaA1 - this will preserve the relative formula
+                    if (cell2.HasFormula && !string.IsNullOrWhiteSpace(cell2.FormulaR1C1)) cell2.ClearFormulaA1();
```
Fixed

igitur commented on ClosedXML/Excel/Cells/XLCellsCollection.cs:
```diff
@@ -414,5 +424,90 @@ private static void IncrementUsage(Dictionary<int, int> dictionary, Int32 key)
+                        NewRowNumber = kvp.Value,
+                        Cells = cells
+                    };
+                }).ToArray();
```
I still have to think about this. I can't imagine that all the tests pass just by accident if there really are side effects.

jahav replied:
As far as I can tell, the code works correctly.

The only side-effect in the `Select` is modification of cell addresses which doesn't affect rowsCollection directly. 

Maybe the issue was readability? In the code, the `Select`+`ToArray` is used as a `ForEach`. LINQ methods (`Select` among them) should be used as a pure function, i.e. it should produce data based on inputs and not mutate anything (IEnumerable chain can be evaluated multiple times). That is why `ForEach` in most implementations around is a terminal node of IEnumerable chain (returns void instead of IEnumerable) - to make sure evaluation of sideeffects happens only once.

As noted, the loop wouldn't work, even simple switch of two rows 1->2, 2->1 would end up with a lost row).

igitur commented:
@Pankraty I address everything, except the remark at https://github.com/ClosedXML/ClosedXML/pull/1649#discussion_r643129834

igitur commented:
I squashed the 3 "CodeMaid" commits.

jahav committed:
Remove setters from IXLSortElements.

IXLSortElements should be immutale to ensure that the sorting specification won't change after elements have been sorted (it would lead to inconsistencies).

jahav committed:
Separate out parsing logic for getting a range sorting order.

jahav committed:
Fix incorrect sort behavior and improve performance.

Sort didn't take into account types and the sorting was done thus incorrectly. It also used to use unstable sort. It now matches the sorting behavior of Excel (used in Dat - Sort and AutoFilter): numers, texts, logical, error and blank. Blank is always at the end, regardless of sorting order (asc/desc). Errors are not sorted.

The performance is improved by two-phase sorting: In first phase, only row indexes are sorted and we get a correct position of each row. That is done using BCL sorting algorithm O(N*log(N) * log(N) worst case because of stable sorting). After that, rows are swapped, so each row is sapped at most once O(N).

jahav committed:
Simplify code by using record readonly struct in ValueSlice.

By using the record struct feature, there is less code and it is also slightly faster (especially because of value comparison). NET likely has optimized parth for record structs.

jahav commented:
The PR had to be completely refactored because of many changes since its inception.

The idea is sound, but there were few problems: 
* The sorting was done on `XLRangeRows`/`XLRangeColumns`. That means each sorted row was stored in range repository, slowing any other manipulations of a sheet (e.g. deletion of ranges).
* Sorting was done without regard for types of values, e.g. TRUE value as a boolean vs TRUE as a string. They are different from sorting POV.

It now matches the sorting behavior of Excel (used in Data - Sort and AutoFilter): numers, texts, logical, error and blank. Blank is always at the end, regardless of sorting order (asc/desc). Errors are not sorted.

The performance is improved by two-phase sorting: In first phase, only row numbers are sorted (it checks data directly from worksheet) and we get a correct sorted position of each row. That is done using BCL stable sorting algorithm `O(N*log(N) * log(N)` worst case because of stable sorting). After that, rows are swapped, so each row is swapped at most once `O(N)`.

Performance gains can be considerable, e.g. `SortIsFast` test takes 124 seconds  (in develop) vs 0.257 (PR), NET7. I tested some other scenarios and although speedups are not always so dramatic, it is several times faster.

jahav committed:
Add info about breaking change to migration doc.
